src/routes/(blog-article)/containerizing-rust-applications-best-practices/+page.md

```diff
@@ -0,0 +1,1049 @@
+
+### Use The `docker run --user` Argument
+
+You can also overwrite the user running the container with the argument `--user`:
+
+<CodeBlock lang="console">
+
+```console
+docker run --user www-data --rm -it docker-rust-app-running-with-root whoami
+www-data
+```
+
+</CodeBlock>
+
+In this example, even though the image is execute as root by default, it will be
+executed as the `www-data` user.
+
+Notice you can even use a non-existing user in both the host and the docker image.
+
+<CodeBlock lang="console">
+
+```console
+$ docker run --user 1001 --rm -it docker-rust-app-running-with-root bash
+I have no name!@895b0f6a3dbb:/app$ 
+```
+
+</CodeBlock>
+
+All this solutions work but they all have a drawback: you need to know the user
+ID at build time (when you build the docker image). You usually want to run the
+container in different environments and sometimes you want to use a different
+user ID for each environment. For example:
+
+- **For development**: If you are using Ubuntu, your user ID is probably `1000`. When
+you run the container locally you want to run it using that ID, so that you don't
+have any problems with permissions.
+- **For CI**: The servers you are using for continuous integration (for instance, GitHub runners)
+might use an specif user. You could use some cache folders and maybe you need to
+use the same user ID as the CI server.
+- **For production**: You could create an specific user for your application and use that
+user ID.
+
+With the proposed solutions you would need to rebuild the docker image so that the
+user ID inside the container is the same as the host user ID.
+
+### Create The User At Runtime
+
+There is al alternative to the previous solutions that make it possible to **run the
+container with different user IDs without rebuilding the image**.
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+
+## Compile su-exec
+FROM docker.io/library/gcc:bookworm as gcc
+COPY ./contrib/dev-tools/su-exec/ /usr/local/src/su-exec/
+RUN cc -Wall -Werror -g /usr/local/src/su-exec/su-exec.c -o /usr/local/bin/su-exec; chmod +x /usr/local/bin/su-exec
+
+## Application
+FROM rust:bookworm as builder
+WORKDIR /app
+RUN cargo init
+RUN cargo build --release
+CMD ["./target/release/app"]
+
+## Runtime
+FROM gcr.io/distroless/cc-debian12:debug as runtime
+RUN ["/busybox/cp", "-sp", "/busybox/sh","/busybox/cat","/busybox/ls","/busybox/env", "/bin/"]
+
+COPY --from=builder /app/target/release/app /usr/local/bin/app
+COPY --from=gcc --chmod=0555 /usr/local/bin/su-exec /bin/su-exec
+COPY --chmod=0555 ./share/container/entry_script_sh /usr/local/bin/entry.sh
+
+ENTRYPOINT ["/usr/local/bin/entry.sh"]
+CMD ["/usr/local/bin/app"]
+
+```
+
+</CodeBlock>
+
+This is the approach we use in Torrust. You run the docker as `root` but we always
+use an entrypoint. That entrypoint creates a new user with an ID provided as an
+environment variable.
+
+The `entry.sh` script is always called when you run the container because it's
```

Comment:
Just a note on the the the usage of "it's".
In college I was taught that you do not use it in formal writing. In formal writing you would use "it is".
Here though I think that you can use it since it is written in a colloquial manner.